### PR TITLE
Initial work for client credentials flow sample project. 

### PR DESCRIPTION
**Pull request by ash-leigh**
Main features:

- Built using dotnet 6
- Implements simple razor pages to provide a UI
- Implements refit to interface with agency and organisation APIs
- Implements RequestClientCredentialsTokenAsync within a DelegatingHandler thats adds the returned token to headers
- Readme with steps required to run sample locally

### CODE REVIEW

File: samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Pages/Index.cshtml.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+            {
+                var request = new AgencyInviteRequest(Form.EmailAddress, Form.MobileNumber, Form.ScopesList, Form.OrganisationId, agencyOptions.Value.ClientNotifyPath, agencyOptions.Value.BrowserRedirectPath);
+                var response = await agencyInviteService.SendInvite(request, cancellationToken);
+                if(response.IsSuccessStatusCode)
+                {
+                    Submitted = true;
+                }
+                else
+                {
+                    var message = response.ReasonPhrase;
```
Comment (alexszym) [inline samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Pages/Index.cshtml.cs]: Maybe log it instead? Or show error

File: samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Services/Agency/Authentication/AuthHeaderHandler.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    {
+        private readonly IOptions<LenusClientOptions> clientOptions;
+
+        public AuthHeaderHandler(IOptions<LenusClientOptions> clientOptions)
+        {
+            this.clientOptions = clientOptions;
+        }
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            var client = new HttpClient();
```
Comment (alexszym) [inline samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Services/Agency/Authentication/AuthHeaderHandler.cs]: Is this a copy from another sample? If not, you should avoid instantiating HttpClient yourself and use HttpClientFactory instead: https://docs.microsoft.com/en-us/aspnet/core/fundamentals/http-requests?view=aspnetcore-6.0

File: samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Services/Agency/ResponseHandler.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+﻿namespace Lenus.Samples.ClientCredentialsFlow.Services.Agency.Authentication
+{
+    class ResponseHandler : DelegatingHandler
```
Comment (alexszym) [inline samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Services/Agency/ResponseHandler.cs]: This doesn't do anything, is it for debugging?

File: samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Configuration/ConfigureAgencyOptions.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
+﻿namespace Lenus.Samples.ClientCredentialsFlow.Configuration
+{
+    using Microsoft.Extensions.Options;
```
Comment (owen-roberts) [inline samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Configuration/ConfigureAgencyOptions.cs]: Some weird formatting going on here causing usings to be within the namespace. This doesn't seem to be the case for other projects in this repo?

File: samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Configuration/LenusClientOptions.cs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+﻿namespace Lenus.Samples.ClientCredentialsFlow.Configuration
+{
+	public class LenusClientOptions
+	{
+        public string TokenUrl { get; set; } = string.Empty;
+        public string ClientId { get; set; } = string.Empty;
+        public string ClientSecret { get; set; } = string.Empty;
+    }
```
Comment (owen-roberts) [inline samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Configuration/LenusClientOptions.cs]: Again just formatting, there's some odd whitespace going on in this file.

File: samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Services/Agency/ResponseHandler.cs (hunk 3 above)
Comment (owen-roberts) [inline samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Services/Agency/ResponseHandler.cs]: Needs an access modifier e.g. public

File: samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/wwwroot/js/site.js [hunk 6]
```diff
@@ -0,0 +1,4 @@
+﻿// Please see documentation at https://docs.microsoft.com/aspnet/core/client-side/bundling-and-minification
+// for details on configuring this project to bundle and minify static web assets.
+
+// Write your JavaScript code.
```
Comment (owen-roberts) [inline samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/wwwroot/js/site.js:4]: Is this file required? Seems empty

File: samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Services/Agency/ResponseHandler.cs (hunk 3 above)
Comment (owen-roberts) [reply samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Services/Agency/ResponseHandler.cs]:
```suggestion
    public class ResponseHandler : DelegatingHandler
```

File: samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/wwwroot/js/site.js (hunk 6 above)
Comment (Philo) [reply samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/wwwroot/js/site.js:4]: This is part of the default project templates, if removed dont forget to also update the _Layout.cshtml

**Comment by Philo** [comment]
Also ensure that the other readme's within the solution as a whole are updated.  There is a root readme.md with some info in it and and make it clear where a sample is written for .net 3.1 -vs- .net 6.

**Comment by Philo** [comment]
Arguably we could update all samples to .net 6.
cc: @owen-roberts and @alexszym ?

**Review by alexszym — COMMENTED** [review]
I've added some initial comments, I know you're still working on this so some of them may not be relevant :)

File: samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/appsettings.Development.json [hunk 7]
```diff
@@ -0,0 +1,21 @@
+{
+  "Logging": {
+    "LogLevel": {
+      "Default": "Information",
+      "Microsoft.AspNetCore": "Warning"
+    }
+  },
+  "Lenus": {
+    "OpenIdConnect": {
+      "TokenUrl": "https://localhost:54580/connect/token",
```
Comment (Philo) [inline samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/appsettings.Development.json]: The other samples have the sandbox environment URLs setup in here rather than any localhost, given that no one else can run identity locally it would be better to set developers on the path to success here.  See the other samples appsettings.development.json files as examples.

File: samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Pages/Index.cshtml.cs [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+            [Display(Name = "Patient Mobile Number", Description = "(Optional) This is the mobile number to which the agency request invitation SMS is sent", Prompt = "+44111222333444")]
+            public string MobileNumber { get; set; }
+
+            [DataType(DataType.MultilineText)]
+            [Display(Name = "Requested health data scopes", Description = "Specify the read and/or write scopes you want the user to consent access to, e.g. read.step_count, read.heart_rate", Prompt = "read.step_count")]
+            [Required]
+            public string? Scopes { get; set; }
+
+            [DataType("OrganisationSelect")]
+            [Display(Name = "Organisation", Description = "(Optional) If a known organisation reference is supplied then the consent request will be made on behalf of the organisation, otherwise consent is requested only for the individual agent", Prompt = "00000000-0000-0000-0000-000000000000")]
```
Comment (Philo) [inline samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Pages/Index.cshtml.cs]: This prompt line is less accurate when referring to agency requests via client_credentials.  Using client credentials it should only be possible to create agency requests for an organisation that is approved for your client, I would expect any attempt to call the API without an organisation id with client creds to fail.  Maybe confirm that?

File: samples/dotnet/clientcreds/readme.md [hunk 9]
```diff
@@ -2,5 +2,52 @@
...
+```
+dotnet dev-certs https --clean
+dotnet dev-certs https --trust
+```
+
+Run `dotnet dev-certs https --trust` on macOS and Windows or alternatively trust the dev certificate via your browser
+
 ## Required Setup
 
+You will need to set up a client via identity admin configured to use the client_credentials grant type. This can be done via a deployed environment such as test or in a locally running instance.
```
Comment (alexszym) [inline samples/dotnet/clientcreds/readme.md]: These samples are mostly targeted at third party developers, so best to stick to 'sandbox' and not mention test or local alternatives.

File: samples/dotnet/clientcreds/readme.md [hunk 10]
```diff
@@ -2,5 +2,52 @@
...
+
+For this solution only agency_api scope is required.
+
+Once you have set up your client you will need to enable an organisation for the parent product and client you created. The id of this organisation is required when completing the form in this solution to send an agency invite.
```
Comment (alexszym) [inline samples/dotnet/clientcreds/readme.md]: 'you will' should be replaced with 'Lenus will', there is no way for a third part developer to enable the organisations.

@Philo Can we hard code organisationId for this sample?

File: samples/dotnet/clientcreds/readme.md [hunk 11]
```diff
@@ -2,5 +2,52 @@
...
+  "Lenus": {
+    "OpenIdConnect": {
+      "ClientSecret": "<REPLACE SECRET>",
+    }
+  }
+```
+
+## Testing the project
+
+Once the project is running a simple UI will be launched providing inputs for an email, phone number, organisation id and requested scopes. Submitting this form will make a request to send agency invite via identity platform.
```
Comment (alexszym) [inline samples/dotnet/clientcreds/readme.md]: It might be that we say here: "Contact Lenus support for a valid organisation Id to use".

File: samples/dotnet/clientcreds/readme.md [hunk 12]
```diff
@@ -2,5 +2,52 @@
 
 Sample implementation of the Lenus Health platform demonstrating use client_credentials grant type flow to support requesting agency over a patient by a clinician within an organisation.
 
+## Running locally
```
Comment (alexszym) [inline samples/dotnet/clientcreds/readme.md:5]: It might controversial, but I'd go with less detail here. With this being a c# sample I would expect that the knowledge on how to run a C# sample code is a given, including the certs. I'd maybe mention that the need to have .NET 6 SDK installed, unless it's already mentioned elsewhere

**Review by alexszym — COMMENTED** [review]
A few notes from my side about rewording towards third party devs

File: samples/dotnet/clientcreds/readme.md (hunk 10 above)
Comment (Philo) [reply samples/dotnet/clientcreds/readme.md]: We would need to hard-code client credential secret as it is a combination of `product` -> `client` -> `enabled orgs`.

We need these docs to be written from the point of view of Lenus providing information to a 3rd party, i.e. that these docs allow an external developer to contact Lenus support and say: "Hi we are implementing client creds flow for agency invites and need organisation X enabled for client Y".

Whilst there is benefit internally to the guidance on how to do this within the admin suite, this isnt something we should document publicly as it isn't an action a developer can perform.

File: samples/dotnet/clientcreds/readme.md (hunk 11 above)
Comment (Philo) [reply samples/dotnet/clientcreds/readme.md]: > It might be that we say here: "Contact Lenus support for a valid organisation Id to use".

Agree, either that or something like:

"You will need to work with our support team to ensure you have a client application that is configured for organisation agency invites.  Please contact support who will be happy to assist."

File: samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Configuration/ConfigureAgencyOptions.cs (hunk 4 above)
Comment (ash-leigh) [reply samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Configuration/ConfigureAgencyOptions.cs]: I have moved all of the using statements outside the namespace

File: samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Services/Agency/ResponseHandler.cs (hunk 3 above)
Comment (ash-leigh) [reply samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Services/Agency/ResponseHandler.cs]: This was a class I ended up not using so have removed the file

File: samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Configuration/LenusClientOptions.cs (hunk 5 above)
Comment (ash-leigh) [reply samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Configuration/LenusClientOptions.cs]: Fixed the whitespace issue

File: samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Pages/Index.cshtml.cs (hunk 1 above)
Comment (ash-leigh) [reply samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Pages/Index.cshtml.cs]: Error is logged instead now

File: samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Services/Agency/Authentication/AuthHeaderHandler.cs (hunk 2 above)
Comment (ash-leigh) [reply samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Services/Agency/Authentication/AuthHeaderHandler.cs]: I have used HttpClientFactory instead

File: samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Services/Agency/ResponseHandler.cs (hunk 3 above)
Comment (ash-leigh) [reply samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Services/Agency/ResponseHandler.cs]: it was in preparation for handling the responses but it was really required in the end

File: samples/dotnet/clientcreds/readme.md (hunk 11 above)
Comment (ash-leigh) [reply samples/dotnet/clientcreds/readme.md]: @alexszym @Philo do third party developers have access to mailtrap for sandbox?

Comment (alexszym) [reply samples/dotnet/clientcreds/readme.md]: Sandbox is configured with Sparkpost, not mailtrap. I think mailtrap is only in use on WIP.

Comment (ash-leigh) [reply samples/dotnet/clientcreds/readme.md]: I've updated the readme with that

File: samples/dotnet/clientcreds/readme.md [hunk 13]
```diff
@@ -0,0 +1,47 @@
+# Clinician Agency Flow via client_credentials Sample
+
+Sample implementation of the Lenus Health platform demonstrating use client_credentials grant type flow to support requesting agency over a patient by a clinician within an organisation.
```
Comment (alexszym) [inline samples/dotnet/clientcreds/readme.md]:
```suggestion
Sample implementation demonstrating use of client_credentials grant type flow to support requesting agency over a patient by a clinician within an organisation using the Lenus Health Platform.
```

File: samples/dotnet/clientcreds/readme.md [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+Open solution in IDE of choice and run the Lenus.Samples.ClientCredentialsFlow project. 
+
+## Required Setup
+
+You will need to set up a client via the developer portal configured to use the client_credentials grant type. This can be done via a sandbox using your developer account.
+
+For this solution only agency_api scope is required.
+
+Once you have set up your client you will need contact Lenus support and ask us to enable an organisation for the client you created. The id of this organisation is required when sending an agency invite.
```
Comment (alexszym) [inline samples/dotnet/clientcreds/readme.md]:
```suggestion
Once you have set up your client you will need to contact Lenus support and ask us to enable an organisation for the client you created. The id of this organisation is required when sending an agency invite.
```

**Review by alexszym — CHANGES_REQUESTED** [review]
A few typos, also want to see what this button does on a review 😂

File: samples/dotnet/clientcreds/readme.md [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+  "Lenus": {
+    "OpenIdConnect": {
+      "ClientSecret": "<REPLACE SECRET>",
+    }
+  }
+```
+
+## Testing the project
+
+Once the project is running a simple UI will be launched providing inputs for an email, phone number, and requested scopes. Submitting this form will make a request to send agency invite via the sandbox identity platform, you will see emails come through via sandbox sparkpost.
```
Comment (alexszym) [inline samples/dotnet/clientcreds/readme.md]:
```suggestion
Once the project is running a simple UI will be launched providing inputs for an email, phone number, and requested scopes. Submitting this form will make a request to send agency invite via the sandbox identity platform, you will see invite emails come through.
```

**Review by alexszym — APPROVED** [review]
Looks good!

File: samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/appsettings.Development.json [hunk 16]
```diff
@@ -0,0 +1,19 @@
+{
+  "Logging": {
+    "LogLevel": {
+      "Default": "Information",
+      "Microsoft.AspNetCore": "Warning"
+    }
+  },
+  "Lenus": {
```
Comment (Philo) [inline samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/appsettings.Development.json:8]: @ash-leigh - rather than having these config values blank can we have sandbox values defaulted to assist users in understanding what is expected.

See https://github.com/lenushealth/integration-samples/blob/main/samples/dotnet/clinorg/src/appsettings.Development.json for example

**Review by Philo — CHANGES_REQUESTED** [review]

File: samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Pages/Index.cshtml.cs (hunk 8 above)
Comment (ash-leigh) [reply samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/Pages/Index.cshtml.cs]: This should be redundant now since moving OrganisationId to being fed in from the appsettings instead of the form

File: samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/appsettings.Development.json (hunk 16 above)
Comment (ash-leigh) [reply samples/dotnet/clientcreds/Lenus.Samples.ClientCredentialsFlow/src/appsettings.Development.json:8]: I've added these now @Philo

**Review by Philo — APPROVED** [review]
